pkg/helm/helm.go

```diff
@@ -18,165 +18,163 @@ package helm
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	klog.V(3).Infof("helm client: caching stable repository")
-	cif := c.home.CacheIndex("stable")
-	cr := repo.Entry{
-		Name:  "stable",
-		Cache: cif,
-		URL:   c.repoURL,
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: downloading index file")
 	}
-
-	var settings environment.EnvSettings
-	r, err := repo.NewChartRepository(&cr, getter.All(settings))
+	indexPath, err := c.repositoryClient.DownloadIndex(chartRepo)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	if err := r.DownloadIndexFile(c.home.Cache()); err != nil {
-		return errors.Wrapf(err, "Looks like %q is not a valid chart repository or cannot be reached", cr.URL)
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: loading repository")
+	}
+	indexFile, err := c.repositoryClient.Load(indexPath)
+	if err != nil {
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	f.Update(&cr)
-	f.WriteFile(c.home.RepositoryFile(), 0644)
-
-	// Load the repositories.yaml
-	c.rf, err = repo.LoadRepositoriesFile(c.home.RepositoryFile())
+	chartRepo.IndexFile = indexFile
+	c.SetChartRepo(chartRepo)
 
-	klog.V(5).Infof("helm client: initialized client")
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: successfully initialized")
+	}
 
-	return err
+	return nil
 }
 
-func (c *Client) ListCharts() (map[string]repo.ChartVersions, error) {
-	klog.V(4).Infof("helm client: listing charts")
-
-	charts := map[string]repo.ChartVersions{}
-
-	// TODO: handle non-unique names across repos
-	for _, r := range c.rf.Repositories {
-		n := r.Name
-		f := c.home.CacheIndex(n)
-		index, err := repo.LoadIndexFile(f)
-		if err != nil {
-			return nil, errors.Wrapf(err, "Could not load helm repository index at %s", f)
-		}
+// SetChartRepo allows setting the chart repository being used by the client.
+func (c *Client) SetChartRepo(chartRepo *repo.ChartRepository) {
+	c.chartRepo = chartRepo
+}
 
-		for chart, chartVersions := range index.Entries {
-			charts[chart] = chartVersions
-		}
+// ListCharts lists the charts from the chart repository.
+func (c *Client) ListCharts() map[string]repo.ChartVersions {
+	if l := c.log.V(4).Get(); l != nil {
+		l.Log("helm client: listing charts from %s", c.chartRepo.Config.URL)
+		defer l.Log("helm client: listed charts from %s", c.chartRepo.Config.URL)
 	}
-
-	klog.V(4).Infof("helm client: listed charts")
-
-	return charts, nil
+	return c.chartRepo.IndexFile.Entries
 }
 
+// GetChart gets a chart that exists in the chart repository.
 func (c *Client) GetChart(name, version string) (*repo.ChartVersion, error) {
-	klog.V(4).Infof("helm client: getting chart %s:%s", name, version)
-
-	charts, err := c.ListCharts()
-	if err != nil {
-		return nil, err
+	if l := c.log.V(4).Get(); l != nil {
+		l.Log("helm client: getting chart %s:%s", name, version)
 	}
 
+	charts := c.ListCharts()
```

Comment:
Hmm, why do I get the feeling we should be using [`IndexFile.Get()`](https://pkg.go.dev/helm.sh/helm/v3@v3.2.3/pkg/repo?tab=doc#IndexFile.Get) instead?

Reply:
![Sounds good, doesn't work](https://media.giphy.com/media/26uf2JHNV0Tq3ugkE/giphy.gif)

Reply:
I tried.

Reply:
It doesn't work because we are getting the chart by app version, not by chart version. I expanded this on d63a359.